### PR TITLE
Week 4 chris aument

### PR DESCRIPTION
**Pull request by csaument**
# Submitting Week 4 Assessment

### ✅ Completed Files

Mark the following as confirmation that the files in this repository are complete.

- [x] JavaScript Coding Challenges
- [x] Ruby Coding Challenges
- [x] Interview Questions

### 📝 Feedback

If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -25,27 +27,154 @@ const colors2 = [
...
   "saffron"
 ]
+
+it("takes in an array, removes the first item from the array and shuffles the remaining content", () => {
+      expect(isherYates(colors1)).toEqual(expect.arrayContaining(["yellow", "blue", "pink", "green"]))
+      expect(isherYates(colors1)).not.toEqual(expect.arrayContaining(["purple"]))
+      expect(isherYates(colors2)).toEqual(expect.arrayContaining(["indigo", "periwinkle", "ochre", "aquamarine", "saffron"]))
+      expect(isherYates(colors2)).not.toEqual(expect.arrayContaining(["chartreuse"]))
+})
+})
```
Comment (elysemontano) [inline code-challenges.test.js:37]: Really nice work!  I like that you are also verifying that the first item is actually removed in your tests.  Initially I was going to comment on the different names for the function, but I kept scrolling and saw you created a second function called isherYates.  It would be cool to create a test for both and check the efficiency in your tests of one approach versus the other if you wanted to stretch this even further.  Excellent job on this problem though!  Love the thorough research and that you tried multiple approaches!

File: code-challenges.test.js [hunk 2]
```diff
@@ -25,27 +27,154 @@ const colors2 = [
...
+// ReferenceError: talleyVotes is not defined
 
 // b) Create the function that makes the test pass.
 
+// Create a function talleyVotes to count the votes
+// Input: an object keys for upVotes and downVotes; values associated with both keys are integers
+// Output: a signed integer indicating the total of votes; negative indicates there were more downVotes
+// Access the key:value pairs to perform simple arithmetic
+// Throws type error if any of the elements are not configured properly
+// Number.isSafeInteger provides Boolean response on whether a value is a safe integer
```
Comment (elysemontano) [inline code-challenges.test.js:122]: Ooo!  Nice work on coming up with some edge cases!  It could be cool to add a test that will check that as well for an added stretch.

File: code-challenges.test.js [hunk 3]
```diff
@@ -25,27 +27,154 @@ const colors2 = [
...
+// Deconstruct the elements from the Set object and return them as values in a new array
+// Individual elements passed as arguments will also function properly, so long as they are not the first element
+// Throws typeError if the first element is not an array
+
+const singleValues = (arr1, arr2, ... arrN) => {
+      if(! Array.isArray(arr1)){
+            throw TypeError('the first element must be an array')
+      }
+      return [... new Set(arr1.concat(arr2, ... arrN))]
+}
```
Comment (elysemontano) [inline code-challenges.test.js:173]: Wow!  Outstanding work!!!  👏

File: interview-questions.md [hunk 4]
```diff
@@ -10,40 +10,102 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
+Researched answer: I started going down a rabbit-hole looking into the relative utility of both paradigms within parallel and distributed computing problems. One table comparing the relative strengths and weakness of these options indicated that OOP does not support parallel programming. This seemed like an interesting and strong assertion. Looking deeper, I think that there is a more nuanced explanation that OOP requires some deliberate organization of a system, allowing subcomponents and their objects to function independently of other subcomponents, whereby both subcomponents can complete their calculations independently. Functional programming seems to be much more commonly used in academic environments, particularly by scientists. This is likely a confluence of perpetual education, system efficiency, and a preference for highly-logical (and sometimes linear) styles of thought.
```
Comment (elysemontano) [inline interview-questions.md:19]: 🤯 WOW!  This may be the most thorough initial answer I have seen!  The way you talk about these concepts are so in-depth and well articulated that it seems like it is coming straight out of documentation somewhere (not implying anything because I am pretty confident this is in your own words).  Outstanding work!

File: interview-questions.md [hunk 5]
```diff
@@ -10,40 +10,102 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+p output
+
+# [1, 2, 3]
+
+```
+
 1. What is a block in Ruby?
 
-Your answer:
+Your answer: Ruby blocks group together operations within a unanamous function.
```
Comment (elysemontano) [inline interview-questions.md:57]: This definitely describes a block, there is a lot of room to expand on this answer though.

File: interview-questions.md [hunk 6]
```diff
@@ -10,40 +10,102 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 Researched answer:
 
+A lot of discussion on extracting values from a Ruby hash seem to be centered around SQL Queries in applications built using Ruby on Rails. In this context, the SQL Queries need keys and values separated so that they can be formatted correctly for an INSERT operation within an SQL database. This is accomplished by calling the .extract method on the hash and passing the desired keys or values:
+
+```ruby
+# Extract key
+my_hash.extract! ('key_name').keys
+
+# Extract value
+my_hash.extract! ('key_name').values
```
Comment (elysemontano) [inline interview-questions.md:97]: 👏 Nicely done!

File: ruby-challenges.rb [hunk 7]
```diff
@@ -3,32 +3,92 @@
 
 # --------------------1) Create a method that takes in a number and determines if the number is even or odd. Use the test variables provided.
 
+## Create a function to determine whether a number is even or odd
+# Input: a number (float or integer)
+# Output: a string concatenating input with a statement of even/oddness
+# Function returns a catchall phrase if the input was not an integer
+
+def odd_ball(num)
+      if num % 2 == 0
```
Comment (elysemontano) [inline ruby-challenges.rb:12]: Ruby has a handy shorthand method for checking if a number is even or odd.  If you are curious, look up even? and .odd? methods

**Review by elysemontano — APPROVED** [review]
This assessment is incredible Chris!  You crushed the logic and pseudocode in both code challenges and ruby challenges, and almost all of your interview section were hire worthy and extremely well articulated and detailed!  Currently demonstrating a competent skill level with the current material.  Left a few comments for you to consider.  Keep up the great work!

File: ruby-challenges.rb (hunk 7 above)
Comment (csaument) [reply ruby-challenges.rb:12]: Yes, I saw those during development of our text-based game. Somewhat simpler solution. Half of Ruby is looking up if there's a built-in method for something. The other half seems to be figuring out how I made a syntactical error.

File: interview-questions.md (hunk 5 above)
Comment (csaument) [reply interview-questions.md:57]: I think by this point in the questions, I didn't really have a firm enough understanding of how a Ruby block was distinct from a code block. I could have discussed both, but opted to dive deeper for my researched answer.

File: code-challenges.test.js (hunk 1 above)
Comment (csaument) [reply code-challenges.test.js:37]: Aaron gave me the idea for checking that the first element was removed while I was listening in on the main channel on Zoom.
